std/eval/_std/sys/thread/EventLoop.hx

```diff
@@ -96,45 +128,79 @@ class EventLoop {
 	}
 
 	function refUnref():Void {
-		if(promisedEventsCount > 0) {
+		if (promisedEventsCount > 0 || (isMainThread && haxe.MainLoop.hasEvents())) {
 			wakeup.ref();
 		} else {
 			wakeup.unref();
 		}
 	}
 
 	public function progress():NextEventTime {
-		//TODO: throw if loop is already running
-		if((handle:Loop).run(NOWAIT)) {
+		if (started) throw "Event loop already started";
+
+		if (handle.run(NOWAIT)) {
 			return AnyTime(null);
 		} else {
 			return Never;
 		}
 	}
 
+	/**
+		Blocks until a new event is added or `timeout` (in seconds) to expires.
+
+		Depending on a target platform this method may also automatically execute arriving
+		events while waiting. However if any event is executed it will stop waiting.
+
+		Returns `true` if more events are expected.
+		Returns `false` if no more events expected.
+
+		Depending on a target platform this method may be non-reentrant. It must
+		not be called from event callbacks.
+	**/
 	public function wait(?timeout:Float):Bool {
-		//TODO: throw if loop is already running
-		if(timeout == null) {
+		if (started) throw "Event loop already started";
```

Comment:
At some point we should start throwing something other than raw strings... There might be a use-case for try-catching a `wait` call and specifically handling this exception, as opposed to some exception that comes from further down the call-stack.

On the other hand, one could just check the `started` state manually.

Reply:
tbh I'm not really sure what this wait is supposed to be used for, as it was broken before this PR (no timeout provided = crash, timeout provided = timeout ignored)

Reply:
I suppose; I don't do that kind of things myself :sweat_smile: